In [45]:
# Imports section
import pandas as pd
import numpy as np
import sklearn

## Part 1. Loading the dataset

In [46]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
df = pd.read_csv("science_data_large.csv")

## Part 2. Splitting the dataset

In [47]:
# Take the pandas dataset and split it into our features (X) and label (y)
from sklearn.model_selection import train_test_split
X = df.iloc[:,:-1]
Y = df.iloc[:,-1]

x_train, x_test, y_train, y_test = train_test_split(X,Y,test_size = 0.10,train_size= 0.90, random_state=42)
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

## Part 3. Perform a Linear Regression

In [48]:
# Use sklearn to train a model on the training set
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score


model = LinearRegression()

model.fit(x_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model



sample_datapoint = pd.DataFrame([[469, 647]], columns=['Temperature °C', 'Mols KCL'])  
y_prediction_sample = model.predict(sample_datapoint)
print("Sample Datapoint: \nThis is the prediction of the output:", round(y_prediction_sample[0],5), "\nActual Output: 624474.2571\n")
#print(x_train.shape)
#print(y_train.shape)
#print(y_train.columns)
#print(x_train.columns)

y_prediction = model.predict(x_train)


r2 = r2_score(y_train,y_prediction)
print("The r squared value is:",round(r2,5))


coefficents = model.coef_
intercepts = model.intercept_

print("\nC1=",round(coefficents[0],5),"C2=",round(coefficents[1],5),"Intercept=",round(intercepts,5))


# Report on the score for that model, in your own words (markdown, not code) explain what the score means

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

Sample Datapoint: 
This is the prediction of the output: 664984.8963 
Actual Output: 624474.2571

The r squared value is: 0.86101

C1= 866.14641 C2= 1032.69507 Intercept= -409391.47958


The Sample data point (First row in the sample)'s size estimation by the scientists, remains close to the estimate from the model. This is good because it shows that the model works decently with an example sample. Additionally the r squared value is 0.86101, meaning that 86% of the variance in the size of the slimes is explained by Temperature and Mols. 0.86 is relatively close to 1 proving the model is a good fit for the data, likely does not over fit the data, and captures most of the relationships between the Temperature, mols, and size.

$x_1 = Temperature (°C),x_2 = Mols  (KCL)$     
$h_{(x)} = -409391.47958 + 866.14641x_1 + 1032.69507x_2$

Sample equation: $E = mc^2$

## Part 4. Use Cross Validation

In [49]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

from sklearn.model_selection import cross_val_score
cv_score = cross_val_score(model,x_train,y_train,cv=10)
cv_scoreT = cross_val_score(model,x_test,y_test,cv=10)
print(cv_score)


print(round(np.mean(cv_score),5))
print(round(np.std(cv_score),5))

# Report on their finding and their significance

[0.85704823 0.86460272 0.8150455  0.82683317 0.8746413  0.90201818
 0.88217921 0.84568669 0.8626558  0.85200682]
[0.8931241  0.80805448 0.78530144 0.87866784 0.81657793 0.86311966
 0.8212513  0.89081167 0.86978917 0.79447833]
0.84212
0.02423


After splitting the data into 10 folds, the mean r squared value is 0.85927 when rounded to 5 decimal places. Within each of the folds the furthest it strays from the mean is roughly 0.05, with the standard deviation of 0.02423. Overall this means that on average the model accounts for 0.85827% of the variance in the feature variables (Temperature and Mols) with relatively low variance. This again being high means that the model does a pretty good job using temperature and mols to measure the slimes size.

## Part 5. Using Polynomial Regression

In [50]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
from sklearn.preprocessing import PolynomialFeatures
poly = PolynomialFeatures(degree=2)

X_train_poly = poly.fit_transform(x_train)
X_test_poly = poly.transform(x_test)


model = LinearRegression()
model.fit(X_train_poly, y_train)


y_pred = model.predict(X_test_poly)



r2 = r2_score(y_test, y_pred)
print(round(model.intercept_,5),round(model.coef_[0],5),round(model.coef_[1],5),round(model.coef_[2],5),round(model.coef_[3],5),round(model.coef_[4],5),r2)


# Report on the metrics and output the resultant equation as you did in Part 3.


2e-05 0.0 12.0 -0.0 0.0 2.0 1.0


$x_1 = Temperature (°C),x_2 = Mols  (KCL)$
$h_{(x)} = 12x_2 + 2x_2^2$

The r squared value is exactly 1.0. This means that the model is likely over-fit for the data set. According to the formula all of the variance depends on the heat of Mols, and itself squared, with coefficients. This is highly unlikely and convincingly invalidates the model or other datasets. 